// payload/dist/express/middleware/errorHandler.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = __importDefault(require("../../utilities/logger"));
const errorHandler_1 = __importDefault(require("./errorHandler"));
const errors_1 = require("../../errors");
const logger = (0, logger_1.default)('payload');
const testError = new errors_1.APIError('test error', 503);
describe('errorHandler', () => {
    const res = generateResponse();
    const next = jest.fn();
    const req = generateRequest();
    it('should send the response with the error', async () => {
        const handler = (0, errorHandler_1.default)(generateConfig({ debug: true }), logger);
        await handler(testError, req, res, next);
        expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ errors: [{ message: 'test error' }] }));
    });
    it('should include stack trace when config debug is on', async () => {
        const handler = (0, errorHandler_1.default)(generateConfig({ debug: true }), logger);
        await handler(testError, req, res, next);
        expect(res.send).toHaveBeenCalledWith(expect.objectContaining({ stack: expect.any(String) }));
    });
    it('should not include stack trace when config debug is not set', async () => {
        const handler = (0, errorHandler_1.default)(generateConfig({ debug: undefined }), logger);
        await handler(testError, req, res, next);
        expect(res.send).toHaveBeenCalledWith(expect.not.objectContaining({ stack: expect.any(String) }));
    });
    it('should not include stack trace when config debug is false', async () => {
        const handler = (0, errorHandler_1.default)(generateConfig({ debug: false }), logger);
        await handler(testError, req, res, next);
        expect(res.send).toHaveBeenCalledWith(expect.not.objectContaining({ stack: expect.any(String) }));
    });
    it('should show the status code when given an error with a code', async () => {
        const handler = (0, errorHandler_1.default)(generateConfig(), logger);
        await handler(testError, req, res, next);
        expect(res.status).toHaveBeenCalledWith(503);
    });
    it('should default to 500 when an error does not have a status code', async () => {
        const handler = (0, errorHandler_1.default)(generateConfig(), logger);
        testError.status = undefined;
        await handler(testError, req, res, next);
        expect(res.status).toHaveBeenCalledWith(500);
    });
    it('should call payload config afterError hook', async () => {
        const afterError = jest.fn();
        const handler = (0, errorHandler_1.default)(generateConfig({
            hooks: { afterError },
        }), logger);
        await handler(testError, req, res, next);
        expect(afterError)
            // eslint-disable-next-line jest/prefer-called-with
            .toHaveBeenCalled();
    });
    it('should call collection config afterError hook', async () => {
        const handler = (0, errorHandler_1.default)(generateConfig(), logger);
        await handler(testError, req, res, next);
        expect(req.collection.config.hooks.afterError)
            // eslint-disable-next-line jest/prefer-called-with
            .toHaveBeenCalled();
    });
});
function generateResponse() {
    const res = {
        status: jest.fn(),
        send: jest.fn(),
    };
    jest.spyOn(res, 'status').mockImplementation().mockReturnValue(res);
    jest.spyOn(res, 'send').mockImplementation().mockReturnValue(res);
    return res;
}
function generateRequest() {
    return {
        collection: {
            config: {
                hooks: {
                    afterError: jest.fn(),
                },
            },
        },
    };
}
function generateConfig(overrides) {
    return {
        debug: false,
        hooks: { afterError: jest.fn() },
        ...overrides,
    };
}
