// payload/dist/preferences/graphql/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_type_json_1 = require("graphql-type-json");
const graphql_1 = require("graphql");
const graphql_scalars_1 = require("graphql-scalars");
function registerPreferences() {
    const { findOne, update, delete: deleteOperation, } = this.operations.preferences;
    const valueType = graphql_type_json_1.GraphQLJSON;
    const preferenceType = new graphql_1.GraphQLObjectType({
        name: 'Preference',
        fields: {
            key: {
                type: (0, graphql_1.GraphQLNonNull)(graphql_1.GraphQLString),
            },
            value: { type: valueType },
            createdAt: { type: new graphql_1.GraphQLNonNull(graphql_scalars_1.DateTimeResolver) },
            updatedAt: { type: new graphql_1.GraphQLNonNull(graphql_scalars_1.DateTimeResolver) },
        },
    });
    this.Query.fields.Preference = {
        type: preferenceType,
        args: {
            key: { type: graphql_1.GraphQLString },
        },
        resolve: (_, { key }, context) => {
            const { user } = context.req;
            return findOne({ key, user, req: context.req });
        },
    };
    this.Mutation.fields.updatePreference = {
        type: preferenceType,
        args: {
            key: { type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString) },
            value: { type: valueType },
        },
        resolve: (_, { key, value }, context) => {
            const { user } = context.req;
            return update({ key, user, req: context.req, value });
        },
    };
    this.Mutation.fields.deletePreference = {
        type: preferenceType,
        args: {
            key: { type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString) },
        },
        resolve: (_, { key }, context) => {
            const { user } = context.req;
            return deleteOperation({ key, user, req: context.req });
        },
    };
}
exports.default = registerPreferences;
