// payload/dist/collections/operations/local/create.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const getFileByPath_1 = __importDefault(require("../../../uploads/getFileByPath"));
async function create(options) {
    var _a, _b;
    const { collection: collectionSlug, depth, locale = (_b = (_a = this === null || this === void 0 ? void 0 : this.config) === null || _a === void 0 ? void 0 : _a.localization) === null || _b === void 0 ? void 0 : _b.defaultLocale, fallbackLocale = null, data, user, overrideAccess = true, disableVerificationEmail, showHiddenFields, filePath, overwriteExistingFiles = false, req, draft, } = options;
    const collection = this.collections[collectionSlug];
    return this.operations.collections.create({
        depth,
        data,
        collection,
        overrideAccess,
        disableVerificationEmail,
        showHiddenFields,
        overwriteExistingFiles,
        draft,
        req: {
            ...req,
            user,
            payloadAPI: 'local',
            locale,
            fallbackLocale,
            payload: this,
            files: {
                file: (0, getFileByPath_1.default)(filePath),
            },
        },
    });
}
exports.default = create;
